hls-graph/src/Development/IDE/Graph/Internal/Action.hs

```diff
@@ -114,13 +114,14 @@ actionFinally a b = do
 apply1 :: (RuleResult key ~ value, ShakeValue key, Typeable value) => key -> Action value
 apply1 k = runIdentity <$> apply (Identity k)
 
+-- todo make the result ordered
 apply :: (Traversable f, RuleResult key ~ value, ShakeValue key, Typeable value) => f key -> Action (f value)
 apply ks = do
     db <- Action $ asks actionDatabase
     stack <- Action $ asks actionStack
     (is, vs) <- liftIO $ build db stack ks
     ref <- Action $ asks actionDeps
-    liftIO $ modifyIORef ref (ResultDeps (fromListKeySet $ toList is) <>)
+    liftIO $ modifyIORef ref (ResultDeps [fromListKeySet $ toList is] <>)
```

Comment:
Now that `<>` is non-commutative, I think we want to insert dependencies at the end of the list as these are the most recent dependencies. Alternatively, we could reverse the list of dependencies when we read it in `refresh`.

We also don't want to insert any dependencies if they are already in the list (i.e. in a previous `KeySet`). Perhaps we need an alternative data-structure to be able to do this efficiently. 

I can't think of anything off the top of my head that satisfies all these requirements:

1. Ordered, so that it keeps track of insertion time - we want to build the first dependencies first
2. Efficient insertion
3. Efficient lookup to detect if keys have already been inserted

Perhaps something around an `IntMap KeySet` could work, where the keys in the `IntMap` keep track of insertion time? Maybe we also need to keep a cumulative `KeySet` for property 3.

Reply:
We also want `modifyIORef'` here.

Reply:
For property 3, we could also just maintain state in `refresh` to avoid rebuilding dependencies that we have already built.

Reply:
> For property 3, we could also just maintain state in refresh to avoid rebuilding dependencies that we have already built.

Yep, maybe we can accumulative the visited KeySet in the refresh to avoid duplicated calls.